dns_sd_avahi.c

```diff
@@ -61,82 +211,279 @@ static void __avahi_browser_cb(AvahiServiceBrowser *browser,
+			/* getaddrinfo() returns a list of address structures */
+			if (*ret) {
+				DEBUG("Unable to find host ('%s'): %s\n",
+						ndata->hostname,
+						gai_strerror(*ret));
+				ddata = remove_node(ddata, i);
+			} else {
+				for (rp = res; rp != NULL; rp = rp->ai_next) {
+					fd = create_socket(res,
+							DEFAULT_TIMEOUT_MS);
+					if (fd < 0) {
+						DEBUG("Unable to create %s%s socket ('%s:%d' %s)\n",
+								res->ai_family == AF_INET ? "ipv4" : "",
+								res->ai_family == AF_INET6? "ipv6" : "",
+								ndata->hostname, ndata->port, ndata->addr_str);
+						ddata = remove_node(ddata, i);
+					} else {
+						close(fd);
+						DEBUG("Something %s%s at '%s:%d' %s)\n",
+								res->ai_family == AF_INET ? "ipv4" : "",
+								res->ai_family == AF_INET6? "ipv6" : "",
+								ndata->hostname, ndata->port, ndata->addr_str);
+						i++;
+					}
+				}
+			}
+			freeaddrinfo(res);
+		}
+		ddata = remove_dup_discovery_data(ddata);
+	}
 
 	avahi_service_browser_free(browser);
 err_free_client:
 	avahi_client_free(client);
 err_free_poll:
-	avahi_simple_poll_free(poll);
-	return -ret; /* we want a negative error code */
+	avahi_simple_poll_free(ddata->poll);
+	return ddata;
+}
+
+int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+{
+	struct avahi_discovery_data *ddata;
+	int ret = 0;
+
+	ddata = dnssd_find_hosts(&ret);
+
+	if (ddata) {
+		*port = ddata->port;
+		strncpy(addr_str, ddata->addr_str, addr_len);
+	}
+
+	free_all_discovery_data(ddata);
+
+	return ret; /* we want a negative error code */
+}
+
+struct iio_scan_backend_context {
+	struct addrinfo *res;
+};
+
+static int dnssd_fill_context_info(struct iio_context_info *info,
+		char *hostname, char *addr_str, int port)
+{
+	struct iio_context *ctx;
+	char uri[HOST_NAME_MAX + 3];
+	char description[255], *p;
+	const char *hw_model, *serial;
+	int i;
+
+	ctx = network_create_context(addr_str);
+	if (!ctx) {
+		ERROR("No context at %s\n", addr_str);
+		return -ENOMEM;
+	}
+
+	if (port == IIOD_PORT)
+		sprintf(uri, "ip:%s", hostname);
+	else
+		sprintf(uri, "ip:%s:%d", hostname, port);
+
+	hw_model = iio_context_get_attr_value(ctx, "hw_model");
+	serial = iio_context_get_attr_value(ctx, "hw_serial");
+
+	if (hw_model && serial) {
+		sprintf(description, "%s (%s), serial=%s",
+				addr_str, hw_model, serial);
+	} else if (hw_model) {
+		sprintf(description, "%s %s", addr_str, hw_model);
+	} else if (serial) {
+		sprintf(description, "%s %s", addr_str, serial);
+	} else if (ctx->nb_devices == 0) {
+		sprintf(description, "%s", ctx->description);
+	} else {
+		sprintf(description, "%s (", addr_str);
+		p = description + strlen(description);
+		for (i = 0; i < ctx->nb_devices - 1; i++) {
+			sprintf(p, "%s,",  ctx->devices[i]->name);
+			p += strlen(p);
+		}
+		p--;
+		*p = ')';
+	}
+
+	iio_context_destroy(ctx);
+
+	info->uri = iio_strdup(uri);
+	if (!info->uri)
+		return -ENOMEM;
+	info->description = iio_strdup(description);
+	if (!info->description)
+		return -ENOMEM;
```

Comment:
`free(info->uri)` here?

Reply:
yes - same problem exists in usb.c

Reply:
at some point we may have to look at running valgrind or some other tool to catch leaks;
or maybe writing some mcheck code
http://man7.org/linux/man-pages/man3/mcheck.3.html

Reply:
we do run coverity on this - https://scan.coverity.com/projects/analogdevicesinc-libiio

it should detect these sorts of things as well...

Reply:
i think here the issue may be that we're calling `iio_strdup()`;
no idea how coverity works, but cppcheck [for example] needs to know that the return value of `iio_strdup()` allocates memory/resource;

that way, it can guess/infer which error path needs free-ing;